### PR TITLE
Allow rotation of front, nutrient and ingredient images during editing

### LINKED ISSUE #1860: Allow the user to rotate a picture from product edition
Allow the user to rotate a picture from product edition

### PR DESCRIPTION
**Pull request by KarthikRIyer**
## Description

A new Activity for rotating image has been added. It has been used for front, nutrients and ingredients images.
On clicking on the tick icon the rotated image is saved to disk and its Uri is returned as a result. Then the image is uploaded.

## Related issues and discussion
fixes #1860 
 
 ## Screen-shots, if any
 
whatsapp image 2019-01-12 at 5 32 29 pm

### CODE REVIEW

**Comment by teolemon** [comment]
Can you use the server side rotation as suggested in the bug rather than client side. Otherwise, we're going to end up with loads of duplicate images.
Example:
https://world.openfoodfacts.org/cgi/product_image_crop.pl?code=3266110700910&id=nutrition_fr&imgid=1&angle=90

**Comment by KarthikRIyer** [comment]
I am not very familiar with retrofit, but I'll familiarise myself and complete it ASAP.

**Comment by KarthikRIyer** [comment]
I was testing the API using postman. What exactly does product_image_crop.pl do? Doesn't it rotate the image in its place? I noticed that I'm getting a new display url every time.

**Comment by teolemon** [comment]
You should have the image id in https://world.openfoodfacts.org/api/v0/product/3266110700910.json?fields=images
product_image_crop.pl rotates the image on the server compared to the initial position of the raw image. So you basically have the initial image to download, let the user rotate it as he/she see fits, and then send back the rotation using the URL

https://world.openfoodfacts.org/cgi/product_image_crop.pl?code=3266110700910&id=nutrition_fr&imgid=1&angle=180

**Comment by KarthikRIyer** [comment]
Can someone please review this PR?

**Comment by KarthikRIyer** [comment]
@PrajwalM2212 can you please review this PR?

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/AddProductIngredientsFragment.java [hunk 1]
```diff
@@ -270,21 +271,25 @@ public void onDetach() {
...
+            Intent intent = new Intent(getActivity(), FullScreenImageRotate.class);
             Bundle bundle = new Bundle();
             if (edit_product && !newImageSelected) {
                 bundle.putString("imageurl", imagePath);
+                bundle.putString("code", product.getCode());
+                bundle.putString("id", "ingredients_en");
             } else {
                 bundle.putString("imageurl", "file://" + imagePath);
+                bundle.putString("code", product.getCode());
+                bundle.putString("id", "ingredients_en");
```
Comment (ripetrescu) [inline app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/AddProductIngredientsFragment.java]: Please remove redundant code, and change it to something like:
`
            Intent intent = new Intent(getActivity(), FullScreenImageRotate.class);
            Bundle bundle = new Bundle();
            if (edit_product && !newImageSelected) {
                bundle.putString("imageurl", imagePath);
            } else {
                bundle.putString("imageurl", "file://" + imagePath);
            }
            bundle.putString("code", product.getCode());
            bundle.putString("id", "ingredients_en");
`

**Comment by teolemon** [comment]
Tiny merge conflict to fix

**Comment by teolemon** [comment]
```androidapp/app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java:44: error: cannot find symbol
import openfoodfacts.github.scrachx.openfood.models.Ingredient;
                                                   ^
  symbol:   class Ingredient
  location: package openfoodfacts.github.scrachx.openfood.models
/home/travis/build/openfoodfacts/openfoodfacts-androidapp/app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java:583: error: cannot find symbol
        void onIngredientListResponse(boolean value, ArrayList<ProductIngredient> productIngredients);
                                                               ^
  symbol:   class ProductIngredient
  location: interface OnIngredientListCallback
2 errors
 FAILED
```

**Comment by teolemon** [comment]
I managed to rotate properly :-)
Nice work. However:
```019-01-31 20:43:32.892 6466-6466/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: openfoodfacts.github.scrachx.openfood, PID: 6466
    java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String com.fasterxml.jackson.databind.JsonNode.toString()' on a null object reference
        at openfoodfacts.github.scrachx.openfood.network.OpenFoodAPIClient$11.onResponse(OpenFoodAPIClient.java:427)
        at retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall$1$1.run(ExecutorCallAdapterFactory.java:70)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6718)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
2019-01-31 20:43:32.954 914-1297/? E/InputDispatcher: channel 'c215a87 openfoodfacts.github.scrachx.openfood/openfoodfacts.github.scrachx.openfood.views.MainActivity (server)' ~ Channel is unrecoverably broken and will be disposed!
2019-01-31 20:43:37.257 7222-7222/openfoodfacts.github.scrachx.openfood E/AndroidRuntime: FATAL EXCEPTION: main
    Process: openfoodfacts.github.scrachx.openfood, PID: 7222
    java.lang.IllegalStateException: Fatal Exception thrown on Scheduler.
        at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:111)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6718)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.widget.EditText.setText(java.lang.CharSequence)' on a null object reference
        at openfoodfacts.github.scrachx.openfood.fragments.AddProductIngredientsFragment.setIngredients(AddProductIngredientsFragment.java:497)
        at openfoodfacts.github.scrachx.openfood.views.AddProductActivity.setIngredients(AddProductActivity.java:1288)
        at openfoodfacts.github.scrachx.openfood.fragments.AddProductOverviewFragment$3.onSuccess(AddProductOverviewFragment.java:620)
        at openfoodfacts.github.scrachx.openfood.fragments.AddProductOverviewFragment$3.onSuccess(AddProductOverviewFragment.java:607)
        at io.reactivex.internal.operators.single.SingleObserveOn$ObserveOnSingleObserver.run(SingleObserveOn.java:81)
        at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:109)
        at android.os.Handler.handleCallback(Handler.java:873) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.app.ActivityThread.main(ActivityThread.java:6718) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
2019-01-31 20:43:39.172 914-980/? E/AutofillSession: No assist structure - app might have crashed providing it
```

**Comment by teolemon** [comment]
(The crash happened when rotating 3 or 4 times)
Also, can you double the size of the rotate and validate button. I have big fingers, and they are very hard to reach ;-)

**Comment by teolemon** [comment]
```2019-01-31 20:48:16.776 8613-8613/openfoodfacts.github.scrachx.openfood E/AndroidRuntime: FATAL EXCEPTION: main
    Process: openfoodfacts.github.scrachx.openfood, PID: 8613
    java.lang.IllegalStateException: Fatal Exception thrown on Scheduler.
        at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:111)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6718)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.widget.EditText.setText(java.lang.CharSequence)' on a null object reference
        at openfoodfacts.github.scrachx.openfood.fragments.AddProductIngredientsFragment.setIngredients(AddProductIngredientsFragment.java:497)
        at openfoodfacts.github.scrachx.openfood.views.AddProductActivity.setIngredients(AddProductActivity.java:1288)
        at openfoodfacts.github.scrachx.openfood.fragments.AddProductOverviewFragment$3.onSuccess(AddProductOverviewFragment.java:620)
        at openfoodfacts.github.scrachx.openfood.fragments.AddProductOverviewFragment$3.onSuccess(AddProductOverviewFragment.java:607)
        at io.reactivex.internal.operators.single.SingleObserveOn$ObserveOnSingleObserver.run(SingleObserveOn.java:81)
        at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:109)
        at android.os.Handler.handleCallback(Handler.java:873) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.app.ActivityThread.main(ActivityThread.java:6718) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
``` 
This time, only one rotation and an attempt to save.

**Comment by teolemon** [comment]
@KarthikRIyer have you had a chance to look into the NPE ?

**Comment by KarthikRIyer** [comment]
@teolemon I'm really sorry, but I've had a busy time at college and I have exams coming. I'll try to look into it as soon as possible.

**Comment by teolemon** [comment]
no worries @KarthikRIyer :-)

**Comment by Karljoones** [comment]
@KarthikRIyer any progress on this? :)

**Comment by teolemon** [comment]
@KarthikRIyer @deniger I've rebased the branch. There's still this Exception when I rotate.

**Comment by teolemon** [comment]
Ok, I'm merging and we'll fix the exception a bit later. Thanks @KarthikRIyer :-)

**Pull request merged**